projects/bootloader/scripts/can_datagram.py

```diff
@@ -0,0 +1,343 @@
+
+    def __init__(self, bustype="socketcan", channel=DEFAULT_CHANNEL,
+                 bitrate=CAN_BITRATE, receive_own_messages=False):
+        print("Initializing CAN Bus...")
+        self.bus = can.interface.Bus(
+            bustype=bustype,
+            channel=channel,
+            bitrate=bitrate,
+            receive_own_messages=receive_own_messages)
+
+    def send(self, message):
+        """This sends the Datagrams."""
+
+        assert isinstance(message, Datagram)
+
+        chunk_messages = self._chunkify(message.serialize(), 8)
+        can_messages = []
+        message_arbitration_id = 0b00000010000
+        message_extended_arbitration = False
+
+        # Populate an array with the can message from the library
+        for chunk_message in chunk_messages:
+            can_messages.append(can.Message(arbitration_id=message_arbitration_id,
+                                            data=chunk_message,
+                                            is_extended_id=message_extended_arbitration))
+
+            # After the first message, set the arbitration ID to 0
+            message_arbitration_id = 0b00000000000
+
+        # Send the messages
+        try:
+            for msg in can_messages:
+                self.bus.send(msg)
+            print("{} messages were sent on {}".format(len(can_messages), self.bus.channel_info))
+
+        except can.CanError:
+            print("Message could not be sent")
+
+    def get_bus(self):
+        """Accessor for the bus"""
+        return self.bus
+
+    def _chunkify(self, data, size):
+        """This chunks up the datagram bytearray for easy iteration."""
+        return (data[pos:pos + size] for pos in range(0, len(data), size))
+
+
+class DatagramListener(can.BufferedReader):
+    """This class handles a callback when a datagram is received."""
+
+    def __init__(self, callback):
+        """This registers the callback."""
+        assert callable(callback)
+        self.callback = callback
+        self.datagram_messages = []
+        self.receiving_datagram = False
+        self.num_node_id = -1
+        self.incomplete_data_bytes = False
+        self.num_data_bytes = -1
+
+        super().__init__()
+
+    def on_message_received(self, msg):
+        """This (SHOULD) wait for the first message in the datagram 0x010."""
+        super().on_message_received(msg)
+
+        if(msg.arbitration_id == 0x0010 and not self.receiving_datagram):
+            # Every time we receive the first message, we reset the variables.
+            self.datagram_messages = []
+            self.receiving_datagram = False
+            self.num_node_id = -1
+            self.incomplete_data_bytes = False
+            self.num_data_bytes = -1
+
+            # Here we want to handle what happens when we get the first message in a datagram
+            first_bytearray = msg.data
+
+            # Get the number of node ids
+            self.num_node_id = first_bytearray[HEADER_SIZE]
```

Comment:
Mitchell's implementation sends an empty start message at the beginning to simplify things... I don't think you can assume that every message has 8 bytes. I haven't studied it in detail, but I think your logic in this method might be a bit overcomplicated by trying to deal with messages individually. Really your goal in this method is to answer the question "is there another message in this datagram?", which I think you might be able to do like this:
* every time you get a message (whether it's a start message or a non-start message), add its bytes to a list (I think it's `datagram_messages` here)
* to check if you've received all the messages, check if you have enough bytes to parse the number of nodes, parse the number of nodes if you do, then check if you have enough bytes to parse the data size, parse it if you do, and then check if you've got all the data based on that

(so really the goal is to get enough bytes to extract all the fields that determine the size of the datagram, and leave the rest of the parsing to `deseralize`)